src/aws-api-mcp-server/awslabs/aws_api_mcp_server/server.py

```
@@ -202,17 +206,39 @@ async def call_aws(
-                'Execution of this operation is not allowed because read only mode is enabled. '
-                f'It can be disabled by setting the {READ_ONLY_KEY} environment variable to False.'
-            )
-            await ctx.error(error_message)
-            return AwsApiMcpServerErrorResponse(
-                detail=error_message,
-            )
+        if READ_OPERATIONS_INDEX is None or not is_operation_read_only(ir, READ_OPERATIONS_INDEX):
+            if READ_OPERATIONS_ONLY_MODE:
+                error_message = (
+                    'Execution of this operation is not allowed because read only mode is enabled. '
+                    f'It can be disabled by setting the {READ_ONLY_KEY} environment variable to False.'
+                )
+                await ctx.error(error_message)
+                return AwsApiMcpServerErrorResponse(
+                    detail=error_message,
+                )
+            elif not BYPASS_TOOL_CONSENT:
+                try:
+                    elicitation_result = await ctx.elicit(
+                        message=f"The CLI command '{cli_command}' requires explicit consent. Do you approve the execution of this command?",
+                        schema=Consent,
+                    )
+
+                    if (
+                        not isinstance(elicitation_result, AcceptedElicitation)
+                        or not elicitation_result.data.answer
+                    ):
+                        error_message = 'User rejected the execution of the command.'
+                        await ctx.error(error_message)
+                        return AwsApiMcpServerErrorResponse(
+                            detail=error_message,
+                        )
+                except McpError as e:
+                    if e.error.code == METHOD_NOT_FOUND:
+                        error_message = 'Client does not support elicitation. Use a different client or update the server configuration.'
```

Should we throw an error if the client doesn't support elicitation? That can lead to a bad user experience.

---

Sounds good, let's confirm the experience with a client that doesn't have the support for elicitation so we can make a call on how it would surface the customers from a UX perspective.

---

The error is shown to the LLM, it might just stop but it can also start using bash tools if available to work around this guardrail. It's not perfect but not sure what else we could do.